### DISCUSSION TITLE
Get the app in APIRouter without circular imports

### DISCUSSION THREAD

**Discussion by Bluenix2** [fastapi/fastapi#6860] [Questions] [answered]
### Core issue

See the following example:

```python
import my_router

app = FastAPI()

app.db = ...

app.include_router(my_router.api)
```

```python
from my_app import app

api = APIRouter()

app.db
```

This obviously causes issues with circular imports. Any good way to fix this?

### Potential fixes

Inspired by discord.py's Cogs we can use classes to inject the app. See the following (this could also be achieved with a decorator):
```python
class Api(APIRouter):
    def __init__(self, app):
        self.app = app

    @APIRouter.get("/")  # staticmethod
    async def root(self):
        return self.app.db.do_something()
```
You then inject app while instancing like this:
```python
import my_router

app = FastAPI()

app.db = ...

app.include_router(my_router.Api(app))
```

Now, depending on how routes are already handled this may or may not be possible. Perhaps a decorator could be added to pass app?

```python
@router.get("/")
@pass_app()
async def root(app):
    ...
```

If this is already possible nicely please disregard. Otherwise I think I would be able to offer my time to get this implemented.

### Summary

Directly importing the app causes circular imports. Using classes or a new form of "pass app" decorator could be beneficial.

**Comment by Cleptomania** [comment, marked as answer]
In general if you have circular imports the question shouldn't be how to make the circular import work, but rather how to not have it appear in the first place.

In your example code it appears that you're trying to set a `db` object on the app variable. This is not really the intended way to do handle things. Let's say for instance you have a database that you need to connect to in the routes, you should provide that database to the routes through FastAPI's dependency injection.

This dynamic is covered at https://fastapi.tiangolo.com/tutorial/sql-databases/ however it doesn't do it with using a multi-file structure like with the APIRouter so I've included a quick example below.

Let's say you have a FastAPI application in `main.py`:

```python
from fastapi import FastAPI

from .api_router import api

app = FastAPI()

app.include_router(api)
```

First create a DB specific module, for SQLAlchemy this would be just creating the engine and session, let's call this file `database.py`:

```python
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

SQLALCHEMY_DATABASE_URL = "sqlite:///./sql_app.db"
# SQLALCHEMY_DATABASE_URL = "postgresql://user:password@postgresserver/db"

engine = create_engine(
    SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False}
)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()
```
So to use this database in your API router, you can create a dependency for it, your `api_router.py` file would look something like:

```python
from typing import Generator

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from .database import SessionLocal

def get_db() -> Generator:
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

api = APIRouter()

@api.get("/")
def my_route(db: Session = Depends(get_db)):
    db.doSomething()
    return
```

You can see some more advanced usage of this technique in the FastAPI full stack template project https://github.com/tiangolo/full-stack-fastapi-postgresql

**Reply by YuriiMotov** [reply]
And, if dependency injection doesn't work for you and you still need to access `app` instance, you can access it via `request` object:

```python
@router.get("/")
async def root(request: Request):
    app = request.app
```

**Comment by LukasJerabek** [comment]
May I also ask? I would like also to expand your example with databases according to https://fastapi.tiangolo.com/advanced/async-sql-databases/

Where would you those @app.event place? That too causes circular imports in my case. If I put that in main.py I need also to to create database object there. But when I import the same object into database.py that would cause circular import.
